### ISSUE TITLE
`trigger` crate

### ISSUE THREAD

**Issue by lann** [fermyon/spin#402]
Spin currently has two built-in trigger types: `HttpTrigger` and `RedisTrigger`. There is also a `TimerTrigger` example and ongoing work towards new cron / "Schedule" (#365) and "CloudEvents" (#394) triggers.

These triggers - especially the two built-in types - share some common logic in how they prepare and invoke their handler components. In order to reduce duplication and future divergence, we should extract that commonality into a new `trigger` crate that exposes trait(s) that can be implemented by individual trigger types.

**Comment by Mossaka** [comment]
I can create a trigger trait. This will help me to work on the CloudEvents trigger feature

**Comment by Mossaka** [comment]
I am thinking about merging the two triggers, HTTP and Redis, to one crate called `spin-trigger`. The `spin-trigger` crate exposes a `Trigger` trait

```rust
#[async_trait]
pub trait Trigger<T> {
    pub async fn new(
        builder: Builder<T>,
        app: Application<CoreComponent>,
        config: TriggerConfig<T>,
    ) -> Result<Self>;
    async fn run(&self) -> Result<()>;
}
```

and changes our we import triggers:
```rust
use spin_trigger::{HttpTrigger, RedisTrigger, CloudEventTrigger};
use spin_engine::{Builder, ExecutionContextConfiguration};

fn main() {
  ...
  let builder = self.prepare_ctx_builder(app.clone()).await?;
  let trigger = HttpTrigger::new(builder, app, self.opts.address, tls).await?;
  trigger.run().await?;

}
```

The benefits of doing this are
1. easy to author a trigger
2. easy to import a trigger
3. author a trigger

Let me know your thoughts @lann @radu-matei

**Comment by radu-matei** [comment]
I agree with having a crate that contains the common logic between the triggers, but not sure with merging our triggers into this crate.
Rather, I would expect the existing triggers to remain separate crates that implement the trait from the new one and import the common logic.

Unless you are thinking about something else with merging the existing triggers?

What do you think?

**Comment by lann** [comment]
I haven't put a _ton_ of thought into this yet, but I was vaguely thinking that the `trigger` crate would consist of one or more traits to define the trigger interface plus some common logic to be shared between (most) triggers. Specifically, I'd be happy if we could move this:
https://github.com/fermyon/spin/blob/ad51cdde94517aa7d6824fdc53b420d993d26389/src/commands/up.rs#L158-L172

Ideally we could also come up with generic versions of some of these (and anything else that makes sense!):
https://github.com/fermyon/spin/blob/ad51cdde94517aa7d6824fdc53b420d993d26389/crates/http/src/lib.rs#L70-L82
https://github.com/fermyon/spin/blob/ad51cdde94517aa7d6824fdc53b420d993d26389/crates/http/src/lib.rs#L225-L234
https://github.com/fermyon/spin/blob/ad51cdde94517aa7d6824fdc53b420d993d26389/crates/http/src/lib.rs#L327-L336

**Comment by lann** [comment]
I'd also be happy if we could better decouple triggers from the `Application` struct, but that might make more sense as part of a config refactor.

**Comment by Mossaka** [comment]
Thanks for the pointers @lann @radu-matei 

I was looking at the http trigger and redis trigger impl and didn't find too many common ground. Their usage of `handle` is totally different, so that's why I was thinking the `Trigger` trait might not do too much.

But @lann 's code blocks make sense. I will make a draft PR today and let you to review it.

**Comment by Mossaka** [comment]
> Ideally we could also come up with generic versions of some of these (and anything else that makes sense!):

I am trying to define `ApplicationConfiguration` to be a generic type over the trigger type, like `ApplicationConfiguration<Trigger>`. In fact, the whole `Application` struct could be `Application<Trigger>` since there could only be either Http Application or Redis Application.  Then I found that I need to convert the enum (from toml) to generic struct... which is pretty difficult to do.

**Comment by lann** [comment]
> which is pretty difficult to do.

No problem! We don't need to boil the ocean all at once here.

**Comment by lann** [comment]
@Mossaka I have time today to take a look at this; if you would like we could work together on supporting what you need for a CloudEvents trigger?

**Comment by lann** [comment]
This could be relevant, depending on timing: #417

**Comment by lann** [comment]
Fixed by #418
